### PR TITLE
refactor(pipeline): resource gc refactoring into provider

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
resource gc refactoring into provider

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda-org.erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=290593&iterationID=1146&pId=290591&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  Optimize that resource gc refactoring into provide           |
| 🇨🇳 中文    |     pipeline的resource gc模块化         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: modules/pipeline/providers/resourcegc/log.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcegc
+
+var (
+	logPrefix = "[resourcegc]: "
+)
+
+func (p *provider) infof(format string, args ...interface{}) {
```
Comment (sfwn) [inline modules/pipeline/providers/resourcegc/log.go:21]: no need this prefix anymore.
`module name` will be added automatically by logrus fields.

File: modules/pipeline/providers/resourcegc/gc.go [hunk 2]
```diff
@@ -117,102 +116,102 @@ func (r *Reconciler) ListenGC(ctx context.Context) {
...
-// 1) 若暂未 wait gc，则直接返回
-// 2) 若正在 wait gc，则 lease 设置一个较大的值 (两天)，防止被 GC
+// DelayGC delay GC
+// 1) if not wait gc，return directly
+// 2) if waiting gc，lease set a longer time (two day)，prevent GC
 //
-// 典型场景：重试失败节点时，namespace 不变，若失败 pipeline 被 GC，会导致新的任务也被 GC，重试失败
-func (r *Reconciler) delayGC(namespace string, pipelineID uint64) {
-	notFound, err := r.js.Notfound(context.Background(), makePipelineGCKey(namespace))
+// example：retry failed-task，namespace unchanging，if namespace be gc，will cause task gc, pipeline failed
```
Comment (sfwn) [inline modules/pipeline/providers/resourcegc/gc.go:137]: `typical scenario:`

**Review by sfwn — CHANGES_REQUESTED** [review]
monitor the process of resourcegc online, and optimize it if necessary.

File: modules/pipeline/providers/resourcegc/log.go (hunk 1 above)
Comment (chengjoey) [reply modules/pipeline/providers/resourcegc/log.go:17]: adjusted

File: modules/pipeline/providers/resourcegc/gc.go (hunk 2 above)
Comment (chengjoey) [reply modules/pipeline/providers/resourcegc/gc.go]: adjusted

File: modules/pipeline/providers/resourcegc/log.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourcegc
+
+func (p *provider) infof(format string, args ...interface{}) {
```
Comment (sfwn) [inline modules/pipeline/providers/resourcegc/log.go:17]: remove this file directly.

File: modules/pipeline/providers/resourcegc/pipeline_gc_compensator.go [hunk 4]
```diff
@@ -12,16 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package reconciler
+package resourcegc
```
Comment (sfwn) [inline modules/pipeline/providers/resourcegc/pipeline_gc_compensator.go:15]: rename file name to compensator

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/resourcegc/pipeline_gc_compensator.go (hunk 4 above)
Comment (chengjoey) [reply modules/pipeline/providers/resourcegc/pipeline_gc_compensator.go]: adjusted

File: modules/pipeline/providers/resourcegc/log.go (hunk 3 above)
Comment (chengjoey) [reply modules/pipeline/providers/resourcegc/log.go]: deleted

File: modules/pipeline/providers/resourcegc/compensator.go [hunk 5]
```diff
@@ -82,27 +80,39 @@ func (r *Reconciler) getNeedGCPipelines(pageNum int, isSnippet bool) ([]spec.Pip
...
 	req.AllSources = true
 	req.IncludeSnippet = isSnippet
-	pipelines, _, _, _, err := r.dbClient.PageListPipelines(req)
+	pipelines, _, _, _, err := p.dbClient.PageListPipelines(req)
 	if err != nil {
 		return nil, 0, fmt.Errorf("failed to compensate pipeline req %v err: %v", req, err)
 	} else {
-		for _, p := range pipelines {
-			if !p.Status.IsEndStatus() {
+		for _, pt := range pipelines {
```
Comment (sfwn) [inline modules/pipeline/providers/resourcegc/compensator.go:87]: strange `pt`.
I think you can use `r` as receiver name (abbr for `resourcegc`).

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/resourcegc/compensator.go (hunk 5 above)
Comment (chengjoey) [reply modules/pipeline/providers/resourcegc/compensator.go]: adjusted

**Review by sfwn — APPROVED** [review]

**Pull request merged**